RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc

```diff
@@ -278,24 +278,24 @@ void PulseShapeFitOOTPileupCorrection::setPUParams(bool   iPedestalConstraint, b
 }
 
 void PulseShapeFitOOTPileupCorrection::setPulseShapeTemplate(const HcalPulseShapes::Shape& ps, bool isHPD) {
-  // initialize for every hit now to avoid incorrect settings for different channel types (HPD vs SiPM)
-  // FIXME: keep this as a reminder to improve and reduce CPU use
-  //  if( cntsetPulseShape ) return;
-
-   // set the M2 parameters before defining the shape
-   setChi2Term(isHPD);
-
-   resetPulseShapeTemplate(ps);
-
+  // initialize for every different channel types (HPD vs SiPM)
+
+  if (!(&ps == currentPulseShape_ && isHPD == isCurrentChannelHPD_))
+    {
+      setChi2Term(isHPD);
+      resetPulseShapeTemplate(ps);
+      currentPulseShape_ = &ps;
+      isCurrentChannelHPD_ = isHPD;
+    }
 }
 
 void PulseShapeFitOOTPileupCorrection::resetPulseShapeTemplate(const HcalPulseShapes::Shape& ps) { 
    ++ cntsetPulseShape;
    psfPtr_.reset(new FitterFuncs::PulseShapeFunctor(ps,pedestalConstraint_,timeConstraint_,addPulseJitter_,applyTimeSlew_,
 								 pulseJitter_,timeMean_,timeSig_,pedMean_,pedSig_,noise_));
-   spfunctor_    = new ROOT::Math::Functor(psfPtr_.get(),&FitterFuncs::PulseShapeFunctor::singlePulseShapeFunc, 3);
-   dpfunctor_    = new ROOT::Math::Functor(psfPtr_.get(),&FitterFuncs::PulseShapeFunctor::doublePulseShapeFunc, 5);
-   tpfunctor_    = new ROOT::Math::Functor(psfPtr_.get(),&FitterFuncs::PulseShapeFunctor::triplePulseShapeFunc, 7);
+   delete spfunctor_; spfunctor_    = new ROOT::Math::Functor(psfPtr_.get(),&FitterFuncs::PulseShapeFunctor::singlePulseShapeFunc, 3);
```

Comment:
If you changed `*functor_` to be `std::unique_ptr<ROOT::Math::Functor>` this would automatically handle deleting both here and in the destructor.